core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java

```diff
@@ -3301,6 +3301,71 @@ private void checkHavingAliasSameAsColumn(boolean upperAlias) {
     sql(query).ok(expected);
   }
 
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-5013">[CALCITE-5013]
+   * Unparse SqlSetOperator should be retained parentheses
+   * when the operand has limit or offset</a>. */
+  @Test void testSetOpRetainParentheses() {
```

Comment:
I'm suspect whether we should remove the parentheses?
If the SQL is :
```
"SELECT \"product_id\" FROM \"product\""
        + "UNION ALL\n"
        + "(SELECT \"product_id\" FROM \"product\")\n"
        + "INTERSECT ALL\n"
        + "(SELECT \"product_id\" FROM \"product\" WHERE \"product_id\" > 10)"
```
```
SELECT \"product_id\"
FROM \"foodmart\".\"product\"
UNION ALL
SELECT *
FROM (SELECT \"product_id\"
FROM \"foodmart\".\"product\"
INTERSECT ALL
SELECT \"product_id\"
FROM \"foodmart\".\"product\"
WHERE \"product_id\" > 10)
```

I think it has become unclear? So maybe we should stay the parentheses?

Reply:
Do you mean to we should retain it when there is a `where` condition? Or we always retain parentheses?

Reply:
I disagree that we should add parentheses for aesthetics. See my comment in https://issues.apache.org/jira/browse/CALCITE-5013 and let's discuss further there.